### ISSUE TITLE
issue with threaded replies and visibility

### ISSUE THREAD

**Issue by hoodiek** [tootsuite/mastodon#297]
tweets belonging to a thread that are replies to a different user, but chained to your own toots (like when you reply to yourself several times) show up on the home timeline. Example: the toots up to this one are invisible on the timeline, and the remaining ones are visible https://mastodon.social/users/hoodie/updates/234259

this one was reported to me by another user seeing my toots to sophia on their home timeline, despite not following sophia

**Comment by hoodiek** [comment]
this seems to be caused by the system of replies only being hidden when they are DIRECTLY to a person, IE: replying to a toot of theirs as well as @ ing them. i think this is one of the flaws with such a system, and i personally thing all toots starting with @ ing a person should be treated as replies.

**Comment by hoodiek** [comment]
how to fix it if you wish to retain current functionality of @ startinga  tweet not being hidden: if toot starts with @ /and/ is a reply to a different toot, hide it

**Comment by Gargron** [comment]
Well, I agree it's a problem, but this ain't as easy as just checking if the first thing in a toot is @. Because other instances/software choose freely how to format mentions. For example, GNU social keeps the @ outside the anchor tag. We keep it inside. Friendica doesn't use @, and both GNU social and Friendica allow mentioning people without actually mentioning them in the text.

**Comment by hoodiek** [comment]
that makes sense. i still think that toots should be hidden if part of a chain of replies, which could be alloted for, albeit with recursive code

**Comment by Gargron** [comment]
Yeah I'm a bit worried about sticking recursive checks there considering it'd be repeated for each follower. Note taken, anyway

**Comment by hoodiek** [comment]
thanks for listening

**Comment by sophia-park** [comment]
The only alternative solution I can think of would be removing replies from the home timeline period, making every reply thread sort of like a permalink-visible Direct Message only visible from notifications. But that's a step too far, and recursive checks would be a bit expensive. Hmm.

I'm happy seeing replies to all users right now, but I'm unsure I would be if this kept scaling to more and more users.

**Comment by hoodiek** [comment]
aha! if i knew you'd had a github, i'd have added you to the conversation. but yes, i think for now it's not incredibly urgent? but i think it should DEFINITELY be looked into later on, cuz the bigger this grows, the more of a problem it will be

**Comment by halcy** [comment]
I'm seeing two problems pop up

1. Replies - even when threaded - from external seem to show up in the public TL. Unsure why this is - are they not properly tagged? Tagged using some different mechanism?
2. Mastodon does not seem to mark toots that start with an @, but are not threaded, as a reply at all, so they do show in public and home up, which seems like very weird behaviour to me...

**Comment by Gargron** [comment]
Two possibilities:

* It's not a reply (reply-to-a-status) but a mention 
* The status it replied to is not in the system. this triggers the thread resolve mechanism, however whether it completes or not, the original status is distributed before that's known aka with an unset in_reply_to_id, therefore even if later it's set as a reply it's already on people's feeds

Notes:

Different software can format the body differently. You can even mention someone without actually mentioning them in the text at all, from Friendica and GNU social. So guessing from whether a status body begins with "@" will not always work.

**Comment by halcy** [comment]
How would you feel about a "Hide mentions" setting that uses essentially a heuristic approach that makes the assumption that if it starts with a reference to somebody who you're not following, you shouldn't see it (excluding @ characters)? I feel that this would produce the result most people expect:

- When you start a status with a mention, it is not meant to be seen by others.
- When the mention is somewhere else in the status, it is still a mention but _is_ in fact meant to be seen by others (compare twitter dot at or putting the mention at the end or even not putting it in the text at all)

It breaks down if some implementation decides that a character other than @ should be used as a username signifier - do people do that?

**Comment by Gargron** [comment]
Friendica does the facebook thing where it's just a full name as a link. GNU social puts the @ outside the link tag. We put the @ inside the link tag.

**Comment by inmysocks** [comment]
Can we make removing @ replies from the public timeline a user setting? I know that there are people who really want the feature but over the past few days I am really feeling the lack of conversation in the public timeline. It is very hard to follow or join conversations unless you follow everyone this way.

**Comment by marrus-sh** [comment]
Is there a way to carry this information within the toot itself, through some sort of metadata? For example, if you reply to your own toot, and that toot is a reply to someone else, setting a flag on the toot which indicates as much before posting. Since the issue only occurs when you're replying to yourself, this flag could be carried through any replies you make and federation isn't as big of a concern. (Of course, this bug will persist for any non-Mastodon people you follow, but I don't know if there's anything which can be done about that.)

**Comment by hoodiek** [comment]
this thread can basically be closed, i believe there are seperate issues for the issue of blocked users. CC @Gargron

**Comment by Gargron** [comment]
Oh yes I believe this is fixed, I forgot about this issue.
